move toggle to app config

mkangia opened the pull request:
building off suggestion [here](https://github.com/dimagi/Vellum/pull/941#issuecomment-391340803)
Still work in progress.

- [x] move toggle to app configuration
- [x] add link to doc ([Added](https://confluence.dimagi.com/display/internal/Hide+languages+in+forms))
- [x] show appt message to user when opening form with > 2 langs
- [x] get product team approval

@orangejenny 
instead of having three possible values, just thought of keeping it an on/off since the behaviour we changed is applicable if there are more than 2 languages. 
Also, you suggested to keep it on by default.
By default i have kept it Off for now so that people don't see a change in the current behaviour. Or you suggest we should still keep it On by default? i am open to both.

Also thought it was important to let people about the change in behaviour,
Depending on what we choose as default, 
Thinking of adding a message at the top when one opens a form that has more than 2 languages. 
[Default: Off] "You can hide multiple languages in the form. learn more about it here" 
Or
[Default: On] "When working in a form, you will only see the currently-selected language. Click here to learn how to change this behaviour."
and then fade away shortly.

<img width="1428" alt="screen shot 2018-05-30 at 4 49 04 pm" src="https://user-images.githubusercontent.com/3864163/40718011-19648b40-642c-11e8-8a39-c4ac84522d8b.png">

mkangia committed:
move toggle to app config

mkangia committed:
lint [skip ci]

orangejenny commented on corehq/apps/app_manager/templates/app_manager/partials/supported-languages.html:
```diff
@@ -38,6 +38,11 @@
+<div data-bind="visible: showSmartLangDisplayOption">
+    <input type="checkbox" data-bind="checked: smartLangDisplay"/>
+    {% trans "Smart language display in forms" %}.
+    <a href="#" target="_blank">Learn more</a>.
```
just flagging that this doc needs to exist

orangejenny replied:
oh, I missed that was already in your checklist, you're on top of things ;)

orangejenny commented:
This looks good. Sounds fine to just have on/off if it's combined with the messaging you described so current users learn about the new setting. In case you haven't seen it, you can do that kind of messaging with the notifications app - it shows the message with an X and once the user closes it they never see it again. Check out 
 `corehq/apps/notifications/ui_notify.py`

In the Application model, `default=False` makes sense, but it might be nice to make the ui nudge users to turn it on - maybe set it to true in `addLanguage` when a second language is added? They'd still have the option to turn it off. I don't feel strongly about this, just a suggestion.

mkangia committed:
if no preference set, then default to true

mkangia committed:
add ability to add custom checks

mkangia committed:
add notification for new setting

mkangia commented:
Thanks @orangejenny for the ui_notify module. Works great.

Read 
> This is still in it's early stages
    and used in App Builder. Will build upon this as we go along.
and made a change to the implementation there to introduce custom checks. let me know what you think about that?

mkangia commented:
@orangejenny 

> In the Application model, default=False makes sense, but it might be nice to make the ui nudge users to turn it on - maybe set it to true in addLanguage when a second language is added?

Cool. That sounds good to me.. made change for that.
also, that should be third language and not second language. Just saying to make sure we are on the same page.

best to review commit/commit for changes.

mkangia committed:
lint

mkangia committed:
lint

orangejenny commented on corehq/apps/notifications/ui_notify.py:
```diff
@@ -101,3 +111,8 @@ def enabled(self, request):
+
+ABILITY_TO_HIDE_TRANSLATIONS = StaticUINotify(
+    'ability_to_hide_translations',
+    custom_checks=[ensure_app_with_multiple_langs]
```
I think it'd be simpler to just not show the bubble in the template if the app doesn't have multiple langs, rather than add complexity to this file. Then you could also avoid attaching the app to the request in `form_source`, which feels a little messy (probably because adding things to the request feels similar to using global variables).

mkangia replied:
That is definitely simpler. Though i liked how all the logic related to the notification stays here with the notification. By having a check in the view separately, breaks up the logic at two different places. But if this looks like overkill, happy to keep it simple as well.

> app to the request

i thought this would be similar to how we have `couch_user` or `domain` on the request object.

orangejenny replied:
Yeah, I don't like attaching `couch_user` and `domain` to the request either ;)  I think our bar for adding more data to the request object should be high, especially when the data is as large as an app.

orangejenny commented:
> should be third language and not second language

oops, yes, you're right.

I should have flagged earlier, but @dimagi/product should also confirm they're okay with this

mkangia committed:
remove custom_checks and add check add view only

mkangia committed:
Merge branch 'master' into add-config-smart-lang

mkangia committed:
show notification only when viewing form

mkangia committed:
add link to doc[skip ci]

mkangia committed:
remove . [skip ci]

mkangia commented:
Thanks for the review @orangejenny 
Updated the logic to keep it simple as discussed in https://github.com/dimagi/commcare-hq/pull/20705#discussion_r191751864. 
Also, adding a related fix in vellum, https://github.com/dimagi/Vellum/pull/943

Also, added a confluence page and added that for the links.
fyi @dimagi/product for inputs.

orangejenny commented on corehq/apps/app_manager/views/apps.py:
```diff
@@ -254,7 +254,10 @@ def _get_setting(setting_type, setting_id):
     context.update({
-        'smart_lang_display_enabled': app.smart_lang_display
+        'smart_lang_display_enabled': app.smart_lang_display,
+        'show_ui_notification_to_hide_translations': (
```
For some reason I had in my head that getting rid of `custom_checks` would allow logic to be only in the template without anything in the context...forgot that django templates can't handle `>`. In any case, I do think it's simpler this way, thank you for updating.

mkangia committed:
update link [skip ci]
